mmdet3d/models/necks/dla_neck.py

```diff
@@ -0,0 +1,192 @@
+
+
+class DLAUp(nn.Module):
+
+    def __init__(self,
+                 startp,
+                 channels,
+                 scales,
+                 in_channels=None,
+                 norm_cfg=nn.BatchNorm2d):
+        super(DLAUp, self).__init__()
+        self.startp = startp
+        if in_channels is None:
+            in_channels = channels
+        self.channels = channels
+        channels = list(channels)
+        scales = np.array(scales, dtype=int)
+        for i in range(len(channels) - 1):
+            j = -i - 2
+            setattr(
+                self, 'ida_{}'.format(i),
+                IDAUp(channels[j], in_channels[j:], scales[j:] // scales[j],
+                      norm_cfg))
+            scales[j + 1:] = scales[j]
+            in_channels[j + 1:] = [channels[j] for _ in channels[j + 1:]]
+
+    def forward(self, layers):
+        out = [layers[-1]]  # start with 32
+        for i in range(len(layers) - self.startp - 1):
+            ida = getattr(self, 'ida_{}'.format(i))
+            ida(layers, len(layers) - i - 2, len(layers))
+            out.insert(0, layers[-1])
+        return out
+
+
+@NECKS.register_module()
+class DLA_Neck(nn.Module):
+    """DLA Neck.
+
+    Args:
+        in_channels (list[int]): Number of input channels per scale.
+        start_level (int): the scale level where upsampling starts. Default: 2.
+        end_level (int): the scale level where upsampling ends. Default: 5.
+        norm_cfg (dict): Config dict for normalization layer. Default: None.
+    """
+
+    def __init__(self,
+                 in_channels=[16, 32, 64, 128, 256, 512],
+                 start_level=2,
+                 end_level=5,
+                 norm_cfg=None):
+        super(DLA_Neck, self).__init__()
+        self.start_level = start_level
+        self.end_level = end_level
+        scales = [2**i for i in range(len(in_channels[self.start_level:]))]
+        self.dla_up = DLAUp(
+            startp=self.start_level,  # 2
+            channels=in_channels[self.start_level:],
+            scales=scales,  # [1, 2, 4, 8]
+            norm_cfg=norm_cfg)
+        self.ida_up = IDAUp(
+            in_channels[self.start_level],
+            in_channels[self.start_level:self.end_level],
+            [2**i for i in range(self.end_level - self.start_level)], norm_cfg)
+
+    def forward(self, x):
+        x = self.dla_up(x)
+        y = []
+        for i in range(self.end_level - self.start_level):
+            y.append(x[i].clone())
+        self.ida_up(y, 0, len(y))
+        return y[-1]
+
+    def init_weights(self, pretrained=True):
+        """Initialize the weights of FPN module."""
+        for m in self.modules():
+            if isinstance(m, (nn.ConvTranspose2d, nn.Conv2d)):
+                nn.init.normal_(m.weight, std=0.001)
+                if m.bias is not None:
+                    nn.init.constant_(m.bias, 0)
+            elif isinstance(m, ModulatedDeformConv2dPack):
+                w = m.weight.data
```

Comment:
The following operation is the same as `fill_up_weights`? We can simplify the code using that function